src/v/iceberg/datatypes.cc

```diff
@@ -0,0 +1,279 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/datatypes.h"
+
+#include <variant>
+
+namespace iceberg {
+
+struct primitive_type_comparison_visitor {
+    template<typename T, typename U>
+    bool operator()(const T&, const U&) const {
+        return false;
+    }
+    bool operator()(const boolean_type&, const boolean_type&) const {
+        return true;
+    }
+    bool operator()(const int_type&, const int_type&) const { return true; }
+    bool operator()(const long_type&, const long_type&) const { return true; }
+    bool operator()(const float_type&, const float_type&) const { return true; }
+    bool operator()(const double_type&, const double_type&) const {
+        return true;
+    }
+    bool operator()(const decimal_type& lhs, const decimal_type& rhs) const {
+        return lhs.precision == rhs.precision && lhs.scale == rhs.scale;
+    }
+    bool operator()(const date_type&, const date_type&) const { return true; }
+    bool operator()(const time_type&, const time_type&) const { return true; }
+    bool operator()(const timestamp_type&, const timestamp_type&) const {
+        return true;
+    }
+    bool operator()(const timestamptz_type&, const timestamptz_type&) const {
+        return true;
+    }
+    bool operator()(const string_type&, const string_type&) const {
+        return true;
+    }
+    bool operator()(const uuid_type&, const uuid_type&) const { return true; }
+    bool operator()(const fixed_type& lhs, const fixed_type& rhs) const {
+        return lhs.length == rhs.length;
+    }
+    bool operator()(const binary_type&, const binary_type&) const {
+        return true;
+    }
+};
+
+bool operator==(const primitive_type& lhs, const primitive_type& rhs) {
+    return std::visit(primitive_type_comparison_visitor{}, lhs, rhs);
+}
+
+struct field_type_comparison_visitor {
+    template<typename T, typename U>
+    bool operator()(const T&, const U&) const {
+        return false;
+    }
+    bool
+    operator()(const primitive_type& lhs, const primitive_type& rhs) const {
+        return lhs == rhs;
+    }
+    bool operator()(const struct_type& lhs, const struct_type& rhs) const {
+        return lhs == rhs;
+    }
+    bool operator()(const list_type& lhs, const list_type& rhs) const {
+        return lhs == rhs;
+    }
+    bool operator()(const map_type& lhs, const map_type& rhs) const {
+        return lhs == rhs;
+    }
+};
+
+bool operator==(const field_type& lhs, const field_type& rhs) {
+    return std::visit(field_type_comparison_visitor{}, lhs, rhs);
+}
+bool operator==(const nested_field& lhs, const nested_field& rhs) {
+    return lhs.id == rhs.id && lhs.required == rhs.required
+           && lhs.name == rhs.name && lhs.type == rhs.type;
```

Comment:
For now I'm just using these to ensure my serialization is correct.

You're right though that schema type equivalence may have a different criteria. I'd argue though that if we want equivalence of just types, it should be a dedicated comparison method, rather than operator==.